### PR TITLE
[WIP] The Big Refactor: reddit-mobile@3.0.0

### PR DESCRIPTION
**Pull request by ajacksified**
# Do not merge until this notice goes away

Goals:

* Make it easy to load data from multiple apis asynchronously on the client, and synchronously on the server
* Reduce the copypasted code relevant to loading data
* Reduce the configuration required per page/componenet
* Reduce as much as possible the amount of code needed for routes and component instantiation / `didMount`
* Set up for complex caching rules
* Nuke dead code
* Break apart methods for easier testability
* Fix bugs caused by state mismatch (aka, nuke globals)

Status:

* [x] Upgrade core infrastructure (horse, horse-react, snoode) to support new route lifecycle
* [x] Index page
* [x] Comments page
* [x] Login page
* [x] Register page
* [x] Error pages
* [x] Search page
* [x] FAQ page
* [x] Inbox
  * [x] all
  * [x] direct messages
  * [x] modmail
  * [x] mentions
  * [x] compose
* [x] Submission page
* [x] User profile pages
  * [x] gild
  * [x] profile
  * [x] activity
* [x] subreddit about
* [x] user saved/hidden
* [x] Implement smart cache layer
* [x] Make sure compact still works everywhere
* [ ] figure out what to do with tracking metadata
* [x] figure out why scroll position resetting is weird occasionally
* [x] performance testing
* [x] error pages
* [x] 'undefined' in subreddit about menu

Then, go-time.

### CODE REVIEW

**Comment by curioussavage** [comment]
did you still want to merge this branch onto mine?

**Comment by ajacksified** [comment]
* [ ] figure out what to do with tracking metadata
* [ ] figure out why scroll position resetting is weird occasionally
* [ ] performance testing

Then, go-time.

File: src/views/pages/userSaved.jsx [hunk 1]
```diff
@@ -91,18 +51,16 @@ class UserSavedPage extends BasePage {
 
     return (
       <div className='user-page user-saved'>
-        { loading }
-
         <div className='container Listing-container' >
           <div className='vertical-spacing-top'>
             { noLinks }
             <ListingList
+              { ...props }
```
Comment (curioussavage) [inline src/views/pages/userSaved.jsx]: Probably better to explicitly pass in app instead of all props here, and update the propTypes

File: src/views/pages/userProfile.jsx [hunk 2]
```diff
@@ -80,62 +33,27 @@ class UserProfilePage extends BasePage {
 
     return (
       <div className="user-page user-profile">
-        { loading }
-
         <TopSubnav
-          user={ user }
+          { ...this.props }
```
Comment (curioussavage) [inline src/views/pages/userProfile.jsx]: it look like this component is only included to show the current users name, does it need this passed in?

File: src/views/pages/search.jsx [hunk 3]
```diff
@@ -18,13 +16,25 @@ class SearchPage extends BasePage {
...
     this._lastQueryKey = null;
+  }
 
-    this.state = {
-      data: props.data || {},
-      subreddits: props.subreddits || {},
-      loaded: !!(props.data && props.data.data),
-    };
+  _loadSearchResults() {
+    this.props.data.get('search').then(function(results) {
```
Comment (curioussavage) [inline src/views/pages/search.jsx]: does babel make maps return promises or did you modify the get method?  I ask because in my browser this doesn't work.

File: src/views/pages/search.jsx [hunk 4]
```diff
@@ -18,13 +16,25 @@ class SearchPage extends BasePage {
   constructor(props) {
     super(props);
 
+    if ((!this.state.data || !this.state.data.search) && props.ctx.query.q) {
```
Comment (curioussavage) [inline src/views/pages/search.jsx]: shouldn't `this.state.data.search` be `this.state.data.get('search')`?

File: src/views/pages/BasePage.jsx [hunk 5]
```diff
@@ -1,10 +1,65 @@
...
+      this.state.data = props.dataCache;
+      this.state.loaded = !!props.dataCache;
+    }
+
+    for (var key of this.props.data.keys()) {
+      this.watch(key);
+    };
+  }
+
+  getData(property, defaultReturn) {
```
Comment (curioussavage) [inline src/views/pages/BasePage.jsx]: is this a TODO? switching calling this instead of just accessing the properties?

Comment (ajacksified) [inline src/views/pages/BasePage.jsx]: Oh, need to nuke this, dead code now.

File: src/views/pages/search.jsx (hunk 4 above)
Comment (ajacksified) [inline src/views/pages/search.jsx]: This is correct, as you noted in your edit.

File: src/views/pages/search.jsx (hunk 3 above)
Comment (ajacksified) [inline src/views/pages/search.jsx]: Well, the map is a map of promises (there is a promise at the `search` key.) Try this:

```
var m = new Map();
m.set('search', Promise.resolve('hi'));
m.get('search').then(function(a){ console.log(a); })
```

File: src/views/components/TopSubnav.jsx [hunk 6]
```diff
@@ -7,23 +6,29 @@ import SortDropdown from '../components/SortDropdown';
 class TopSubnav extends BaseComponent {
   constructor(props) {
     super(props);
+
+    this.state = {
```
Comment (curioussavage) [inline src/views/components/TopSubnav.jsx]: what is the reason for not using props anymore?

File: src/views/components/TopNav.jsx [hunk 7]
```diff
@@ -10,104 +9,93 @@ import SeashellsDropdown from '../components/SeashellsDropdown';
...
+          loaded: true,
+        });
+      }.bind(this));
     }
+  }
 
-    globals().app.on(constants.TOP_NAV_SUBREDDIT_CHANGE, this._changeSubredditName);
-    globals().app.on(constants.SIDE_NAV_TOGGLE, this._onToggle);
+  componentDidMount() {
+    this.loadSubreddit(this.state.data.get('subreddit'));
```
Comment (curioussavage) [inline src/views/components/TopNav.jsx]: I kinda like this way of using handling the promise

File: src/views/components/SideNav.jsx [hunk 8]
```diff
@@ -42,28 +42,37 @@ class SideNav extends BaseComponent {
...
   }
 
   componentDidMount() {
-    globals().app.on(constants.TOP_NAV_HAMBURGER_CLICK, this._toggle);
-    globals().app.on('route:start', this._close);
+    this.props.app.on(constants.TOP_NAV_HAMBURGER_CLICK, this._toggle);
+    this.props.app.on('route:start', this._close);
+
+    if (!this.state.subscriptions) {
+      this.data.get('subscriptions').then(function(s) {
```
Comment (curioussavage) [inline src/views/components/SideNav.jsx]: should this be `this.state.data` or is it really on the instance?

File: src/views/components/Loading.jsx [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import React from 'react';
+import React from 'react/addons';
```
Comment (curioussavage) [inline src/views/components/Loading.jsx]: thanks for catching this

File: src/routes.jsx [hunk 10]
```diff
@@ -329,379 +212,311 @@ function routes(app) {
...
+      },
+      user: props.userName,
     });
 
-    try {
-      var key = 'saved-' + hidden.toString() + stringify(this.query);
+    this.props.data.set('activities', saved.get(savedOpts))
 
-      page = (
+    this.body = function(props) {
```
Comment (curioussavage) [inline src/routes.jsx]: where does body get called?  downside maybe could be it relies on some environment (this.query, querystring)?

**Comment by curioussavage** [comment]
:nail_care:

**Comment by curioussavage** [comment]
got a typeError on the search page after the prefs request got a 503.

```  TypeError: Cannot read property 'request' of undefined
      at MixedInApp.error (/Users/awinn/mobile/horse-react/node_modules/horse/src/app.es6.js:100:12)
      at MixedInApp.error (/Users/awinn/mobile/reddit-mobile/src/app-mixin.es6.js:19:19)
      at Object.callee$2$0$ (/Users/awinn/mobile/horse-react/src/server.es6.js:67:22)
      at tryCatch (/Users/awinn/mobile/reddit-mobile/node_modules/babel/node_modules/babel-core/node_modules/regenerator/runtime.js:61:40)
      at GeneratorFunctionPrototype.invoke [as _invoke] (/Users/awinn/mobile/reddit-mobile/node_modules/babel/node_modules/babel-core/node_modules/regenerator/runtime.js:323:22)
      at GeneratorFunctionPrototype.prototype.(anonymous function) [as throw] (/Users/awinn/mobile/reddit-mobile/node_modules/babel/node_modules/babel-core/node_modules/regenerator/runtime.js:94:21)
      at onRejected (/Users/awinn/mobile/reddit-mobile/node_modules/co/index.js:80:24)
      at run (/Users/awinn/mobile/reddit-mobile/node_modules/babel/node_modules/babel-core/node_modules/core-js/modules/es6.promise.js:89:39)
      at /Users/awinn/mobile/reddit-mobile/node_modules/babel/node_modules/babel-core/node_modules/core-js/modules/es6.promise.js:100:28
      at process._tickCallback (node.js:355:11)
```

File: src/views/components/TopSubnav.jsx (hunk 6 above)
Comment (ajacksified) [inline src/views/components/TopSubnav.jsx]: Ah, leftover from when I was lazyloading user, which shouldn't happen anyway. Nuketown

File: src/views/components/TopNav.jsx (hunk 7 above)
Comment (ajacksified) [inline src/views/components/TopNav.jsx]: Right now, if there's an error, it's ignored; not sure what the best handling is. Maybe the default `reddit` logo?

File: src/views/components/SideNav.jsx (hunk 8 above)
Comment (ajacksified) [inline src/views/components/SideNav.jsx]: should be `props.data`

File: src/routes.jsx (hunk 10 above)
Comment (ajacksified) [inline src/routes.jsx]: As we talked yesterday - `this.body` is called by horse-react (https://github.com/reddit/horse-react/pull/4/files#diff-98733154bec3fdb5d92023c16948061aR65 and https://github.com/reddit/horse-react/pull/4/files#diff-a90113d10f77f7029601b6a8015e66eeR28).

File: src/views/pages/userProfile.jsx (hunk 2 above)
Comment (ajacksified) [inline src/views/pages/userProfile.jsx]: In many of the cases, `app`, `token`, `user`, and several other props all need to be sent, so I took this shortcut. I think the proper way to define what's explicitly used is in the `Component.proptypes` at the bottom of each component, which needs to be updated.

It's worth checking memory usage, but as long as React isn't doing something stupid, and it's sending everything by reference instead of making copies, this shouldn't matter. I'll do some profiling to confirm.

**Comment by ajacksified** [comment]
:haircut:

**Comment by curioussavage** [comment]
2 things from staging:

* subreddit about menu shows on index page
* In side nav clicking on `my subreddits` expanded but no children (haven't been able to reproduce yet)

**Comment by ajacksified** [comment]
@curioussavage could you include test cases, or maybe watch network panel? Could the empty comments be because of a 503?

**Comment by curioussavage** [comment]
looking into the comments

but I am also unable to log out. response is a 503

**Comment by ajacksified** [comment]
@curioussavage fixed logout. I see what you mean on the comments now.

**Comment by ajacksified** [comment]
Whoops, wrong button.

**Comment by ajacksified** [comment]
Renaming branch to 3.0.0 and closing.
